apps/ejabberd/src/mod_register.erl

```diff
@@ -74,278 +74,288 @@ clean_opt({registration_watchers, Watchers}) ->
-		true ->
-		    IQ#iq{type = error,
-			  sub_el = [SubEl, ?ERR_BAD_REQUEST]}
-	    end;
-	get ->
-	    {_IsRegistered, UsernameSubels, QuerySubels} =
-		case From of
-		    #jid{user = User, lserver = Server} ->
-			case ejabberd_auth:is_user_exists(User, Server) of
-			    true ->
-				{true, [#xmlcdata{content = User}],
-				 [#xmlel{name = <<"registered">>}]};
-			    false ->
-				{false, [#xmlcdata{content = User}], []}
-			end;
-		    _ ->
-			{false, [], []}
-		end,
-        TranslatedMsg = translate:translate(
-                          Lang, <<"Choose a username and password to register with this server">>),
-		    IQ#iq{type = result,
-		  sub_el = [#xmlel{name = <<"query">>,
-				   attrs = [{<<"xmlns">>, <<"jabber:iq:register">>}],
-				   children = [#xmlel{name = <<"instructions">>,
-				              children = [#xmlcdata{content = TranslatedMsg}]},
-				       #xmlel{name = <<"username">>,
-					      children = UsernameSubels},
-				       #xmlel{name = <<"password">>}
-				       | QuerySubels]}]}
+%% Clients must register before being able to authenticate.
+process_unauthenticated_iq(Sender, Receiver, #iq{type = set} = Stanza, IPAddr) ->
+    process_iq_set(Sender, Receiver, Stanza, IPAddr);
+process_unauthenticated_iq(Sender, Receiver, #iq{type = get} = Stanza, IPAddr) ->
+    process_iq_get(Sender, Receiver, Stanza, IPAddr).
+
+process_iq(From, To, #iq{type = set} = IQ) ->
+    process_iq_set(From, To, IQ, jlib:jid_tolower(From));
+process_iq(From, To, #iq{type = get} = IQ) ->
+    process_iq_get(From, To, IQ, jlib:jid_tolower(From)).
+
+process_iq_set(From, To, #iq{sub_el = Child} = IQ, Source) ->
+    true = is_query_element(Child),
+    handle_set(IQ, From, To, Source).
+
+handle_set(Stanza, ClientJID, ServerJID, Source) ->
+    #iq{sub_el = Query} = Stanza,
+    case has_only_remove_child(Query) of
+        true ->
+            attempt_cancelation(ClientJID, ServerJID, Stanza);
+        {false, more} ->
+            error_response(Stanza, ?ERR_BAD_REQUEST);
+        {false, absent} ->
+            case has_username_and_password_children(Query) of
+                true ->
+                    Credentials = get_username_and_password_values(Query),
+                    register_or_change_password(Credentials, ClientJID, ServerJID, Stanza, Source);
+                false ->
+                    ignore
+            end
+    end.
+
+has_only_remove_child(#xmlel{children = C} = Q) when length(C) =:= 1 ->
+    case exml_query:path(Q, [{element, <<"remove">>}], absent) of
+        absent ->
+            {false, absent};
+        _ ->
+            true
+    end;
+has_only_remove_child(#xmlel{children = C} = Q) when length(C) > 1 ->
+    case exml_query:path(Q, [{element, <<"remove">>}], absent) of
+        absent ->
+            {false, absent};
+        _ ->
+            {false, more}
+    end.
+
+has_username_and_password_children(Q) ->
+    {absent, absent} =/= get_username_and_password_elements(Q).
```

Comment:
What if we have just password or just username?  That will give  `{absent, absent} =/= {pawel, absent}` or   `{absent, absent} =/= {absent, admin123}`. Both give true, but according to the function name - has_username_**_and_**_password they shouldn't.